Fabric.Authorization.Persistence.SqlServer/Stores/SqlServerPermissionStore.cs

```diff
@@ -0,0 +1,129 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Fabric.Authorization.Domain.Exceptions;
+using Fabric.Authorization.Domain.Models;
+using Fabric.Authorization.Domain.Stores;
+using Fabric.Authorization.Persistence.SqlServer.Mappers;
+using Fabric.Authorization.Persistence.SqlServer.Services;
+using Microsoft.EntityFrameworkCore;
+
+namespace Fabric.Authorization.Persistence.SqlServer.Stores
+{
+    public class SqlServerPermissionStore : IPermissionStore
+    {
+        private readonly IAuthorizationDbContext _authorizationDbContext;
+
+        public SqlServerPermissionStore(IAuthorizationDbContext authorizationDbContext)
+        {
+            _authorizationDbContext = authorizationDbContext;
+        }
+
+        public async Task<Permission> Add(Permission model)
+        {
+            var entity = model.ToEntity();
+
+            _authorizationDbContext.Permissions.Add(entity);
+            await _authorizationDbContext.SaveChangesAsync();
+            return model;
+        }
+
+        public async Task<Permission> Get(Guid id)
+        {
+            var permission = await _authorizationDbContext.Permissions
+                .SingleOrDefaultAsync(p =>
+                    p.PermissionId == id
+                    && !p.IsDeleted);
+
+            if (permission == null)
+            {
+                throw new NotFoundException<Permission>($"Could not find {typeof(Permission).Name} entity with ID {id}");
+            }
+
+            return permission.ToModel();
+        }
+
+        public async Task<IEnumerable<Permission>> GetAll()
+        {
+            var permissions = await _authorizationDbContext.Permissions
+                .Where(p => !p.IsDeleted)
+                .ToArrayAsync();
+
+            return permissions.Select(p => p.ToModel());
+        }
+
+        public async Task Delete(Permission model)
+        {
+            var permission = await _authorizationDbContext.Permissions
+                .SingleOrDefaultAsync(p =>
+                    p.PermissionId == model.Id
+                    && !p.IsDeleted);
+
+            if (permission == null)
+            {
+                throw new NotFoundException<User>($"Could not find {typeof(Permission).Name} entity with ID {model.Id}");
+            }
+
+            permission.IsDeleted = true;
```

Comment:
should the RolePermission and UserPermission entities related to this permission also be marked as deleted?

Reply:
That's a good question. Perhaps the queries that retrieve the RolePermission and UserPermission entities should ensure the corresponding Permission is not deleted?

Reply:
Which queries are you referring to? The Get method in this class?

Reply:
I meant the queries that are in the SqlServerUserStore and SqlServerRoleStore.

Reply:
Ahh ok. I think all Get methods should check that the entity being searched for is not marked deleted. My original question here though relates more to the fact that the UserPermission and RolePermission entities implement ITrackable so shouldnt we be marking them deleted as well?

Reply:
Yeah, you're probably right now that I think about it. I just meant for the User#Get, we only pull in related Permission entities that aren't deleted, but it's probably best to just do like you're doing in the group store and mark all relationship entities as deleted as well.